src/qs8-vadd/gen/qs8-vadd-minmax-hvx-u128.c

```diff
@@ -0,0 +1,182 @@
+// Auto-generated file. Do not edit!
+//   Template: src/qs8-vadd/hvx.c.in
+//   Generator: tools/xngen
+//
+// Copyright 2024 Google LLC
+//
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree.
+
+#include <assert.h>
+
+#include <hvx_hexagon_protos.h>
+#include <hexagon_protos.h>
+#include <hexagon_types.h>
+
+#include <xnnpack/vbinary.h>
+#include <xnnpack/intrinsics-polyfill.h>
+
+
+void xnn_qs8_vadd_minmax_ukernel__hvx_u128(
+    size_t batch,
+    const int8_t* input_a,
+    const int8_t* input_b,
+    int8_t* output,
+    const union xnn_qs8_add_minmax_params params[restrict XNN_MIN_ELEMENTS(1)]) XNN_OOB_READS
+{
+  assert(batch != 0);
+  assert(batch % sizeof(int8_t) == 0);
+  assert(input_a != NULL);
+  assert(input_b != NULL);
+  assert(output != NULL);
+
+  const HVX_Vector vbias = Q6_V_vsplat_R(*((int32_t *) &params->hvx.bias));
+  const HVX_Vector va_multiplier = Q6_V_vsplat_R(*((int32_t *) &params->hvx.a_multiplier));
+  const HVX_Vector vb_multiplier = Q6_V_vsplat_R(*((int32_t *) &params->hvx.b_multiplier));
+  const int32_t vshift = params->hvx.shift;
+  const HVX_Vector voutput_zero_point = Q6_Vh_vsplat_R(*((int16_t *) &params->hvx.output_zero_point));
+  const HVX_Vector voutput_min = Q6_Vb_vsplat_R(*((int8_t *) &params->hvx.output_min));
+  const HVX_Vector voutput_max = Q6_Vb_vsplat_R(*((int8_t *) &params->hvx.output_max));
+  int8_t* ptr_o = output;
+
+  for (; batch >= 128 * sizeof(int8_t); batch -= 128 * sizeof(int8_t)) {
+    HVX_Vector va0 = *((HVX_UVector*)input_a);
+    HVX_Vector vb0 = *((HVX_UVector*)input_b);
+    HVX_Vector va1 = *((HVX_UVector*)(input_a + 32));
+    HVX_Vector vb1 = *((HVX_UVector*)(input_b + 32));
+    HVX_Vector va2 = *((HVX_UVector*)(input_a + 64));
+    HVX_Vector vb2 = *((HVX_UVector*)(input_b + 64));
+    HVX_Vector va3 = *((HVX_UVector*)(input_a + 96));
+    HVX_Vector vb3 = *((HVX_UVector*)(input_b + 96));
+    input_a += 128;
+    input_b += 128;
+
+    // unpack: 8 bit to 16 bit
+    HVX_Vector va0_i16_lo = Q6_V_lo_W(Q6_Wh_vunpack_Vb(va0));
```

Comment:
Is this loading 4 overlapping vectors 32 bytes at a time, and then unpacking only the low part of each one?

That seems really inefficient, and requires the use of unaligned loads even if the pointer were aligned. 

I think this should be possible to do by loading one 128 byte vector, and using both the hi and lo parts of the unpack result, instead of only using the lo part.

Reply:
You are correct :) I will change the code in the way you described.